src/main/java/frc/robot/scoreStates/ShootState.java

```diff
@@ -2,62 +2,139 @@
 
 import com.pathplanner.lib.auto.AutoBuilder;
 import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Rotation2d;
 import edu.wpi.first.math.geometry.Translation2d;
 import edu.wpi.first.units.Units;
 import edu.wpi.first.wpilibj2.command.Command;
 import edu.wpi.first.wpilibj2.command.Commands;
 import frc.robot.Constants;
+import frc.robot.commandGroups.CommandGroups;
+import frc.robot.subsystems.hood.Hood;
+import frc.robot.subsystems.shooter.Shooter;
+import frc.robot.subsystems.shooter.ShooterConstants;
 import frc.robot.swerve.SwerveDrive;
 import java.util.List;
 import java.util.Set;
+import lib.PoseEstimation;
 import lib.Utils;
+import lib.math.interpolation.InterpolatingDouble;
 
 public class ShootState implements ScoreState {
+    private static Shooter shooter;
+    private static Hood hood;
+    private Translation2d speakerPose;
+    private InterpolatingDouble distanceToSpeaker;
+    private Pose2d botPose;
+    private List<Translation2d> optimalPoints;
+    private Translation2d optimalTranslation;
+    private double optimalRotation;
+    private static boolean inBounds;
+
+    private double shooterToSpeakerHeight =
+            ScoreStateConstants.SPEAKER_TARGET_HEIGHT
+                    - ShooterConstants.SHOOTER_HEIGHT.in(Units.Meters);
+
+    public ShootState() {
+        shooter = Shooter.getInstance();
+        hood = Hood.getInstance();
+    }
+
+    private Command rotate() {
+        return SwerveDrive.getInstance()
+                .turnCommand(
+                        optimalRotation, ScoreStateConstants.TURN_TOLERANCE.in(Units.Rotations));
+    }
 
     @Override
-    public Command driveToClosestOptimalPoint() {
-        return Commands.defer(
+    public Command calculateTargets() {
+        return Commands.runOnce(
                 () -> {
-                    Translation2d speakerPose;
-                    Pose2d botPose = SwerveDrive.getInstance().getBotPose();
-                    List<Translation2d> optimalPoints;
+                    botPose = PoseEstimation.getInstance().getEstimatedPose();
                     if (isRed()) {
+                        if (botPose.getX()
+                                < ScoreStateConstants.RED_BOUNDS_MAP.get(botPose.getY()).value) {
+                            inBounds = false;
+                        }
                         speakerPose = ScoreStateConstants.SPEAKER_POSE_RED;
                         optimalPoints = ScoreStateConstants.OPTIMAL_POINTS_SHOOT_RED;
                     } else {
+                        if (botPose.getX()
+                                > ScoreStateConstants.BLUE_BOUNDS_MAP.get(botPose.getY()).value) {
+                            inBounds = false;
+                        }
```

Comment:
1. If I understand correctly, `inBounds` can only be `null` or `false`, so the condition on line 103 might throw a `NullPointerException`.
2. [The odometry's origin should be on the blue alliance side,](https://docs.wpilib.org/en/stable/docs/software/basic-programming/coordinate-system.html#always-blue-origin) so I think you need to flip the `<` and `>`.

Reply:
1. bce5b6fd2ce18ae703a03bd35ab64989b512c176
2. The signs are the correct way I considered this.

Reply:
A few enhancements:
1. Extracted X and Y into variables to reduce code repetition and length -- this enabled the condition to be a single line.
2. Directly setting the boolean value to the result of the boolean check -- no need for `if true then true else false`.
3. Flipping `not <` into `>=` which is more concise (actually Spotless can do this too).
```java
    private void updateInBounds() {
        double poseX = botPose.getX();
        InterpolatingDouble poseY = new InterpolatingDouble(botPose.getY());
        if (isRed()) {
            inBounds = poseX >= ScoreStateConstants.RED_BOUNDS_MAP.get(poseY).value;
        } else {
            inBounds = poseX <= ScoreStateConstants.BLUE_BOUNDS_MAP.get(poseY).value;
        }
    }
```
> The signs are the correct way I considered this.

Yes, you're right. I was confused because usually, you don't have double negation. The previous version was "if it's not `some condition`, then the result is `false`". It's better to avoid these double negations.

Also, `BLUE_BOUNDS_MAP` and `RED_BOUNDS_MAP` should include the "shoot" term.

---
Something else I noticed now is that `calculateTargets()` might be called a long time before the `score()` is executed. This could lead to some issues, for example, if `ShootState`'s `calculateTargets()` is called when we are at the Source, so `inBounds = false`, but `score()` might be executed when we are in the Wing, so `inBounds` should not be `false` anymore.

Reply:
4b19c4462ac4c9d566423a8014459e932c5cd524

Reply:
While invoking `updateInBounds()` in DTOP is a positive step forward, the problem is wider than that. A few examples:
1. When the robot is near our Source, switching to state "shoot" runs `ShootState`'s `prepareSubsystems()`. Because we are still very far from the Speaker, the shooter doesn't warm up and only starts moving when `score()` is called.
2. Expanding upon the previous example, the pose we should DTOP to might change -- initially, it can be decided when we're near the Source, but when `score()` runs we might be on the other side of the field, so there is no reason to drive far to the pose determined initially.
3. After the next scoring state is set to `AmpState`, its `calculateTargets()` function is called, deciding that forward Amping is preferred. While driving to the Amp, the robot rotates, making reversed Amping desired. The robot eventually rotates with no need.
---
Another half-relevant point is that we might not want to run `prepareSubsystems()` immediately on state change. For example, there is no reason to set the elevator's height to the climbing height if we're very far from the Stage, but only when we are closer. While I think the first issue I raised here has a rather simple solution (but I'd like to hear yours first), this one is a bit more complex to solve.

---

_And here it is in ChatGPT's Shakespearean sonnet style…_
> Amidst the realm of DTOP, we find,
> A noble quest, though fraught with deep despair,
> Wherein the robot's path, in truth, is blind,
> And errors lurk, with mischief to ensnare.
> 
> Upon the field, by Source and Speaker's side,
> The 'shoot' state beckons, yet the shooter stalls,
> In distant lands, its speed doth yet abide,
> Till score's call wakes it from its silent halls.
> 
> Oh, shifting pose! Thou art a fickle sprite,
> Decreed near Source, yet found afar at score,
> No need to journey towards the distant light,
> When fate hath placed us on a distant shore.
> 
> Prepare, prepare! But timing is a jest,
> For actions planned may prove unwelcome guests.
> 
> In AmpState's realm, where targets are decreed,
> The forward path is chosen, clear and bright,
> Yet as we move, the path doth twist with speed,
> Reversed we go, despite our guiding light.
> 
> And so, amidst the chaos, we confess,
> The simple fix, and complex riddle, stress.

Reply:
1. So i am removing the only if in the shooter & the hood?
2. Notice that the DTOP happen in the score, so if we switched sides the optimal pose will be recalculated
3. If the robot can amp backwards it will, if it can't we will amp forward, I don't see the problem here
4. We don't care that the elevator gets set immediately this doesn't matter and is not an issue

Reply:
1. No, that's not what I meant. What I suggest is that `prepareSubsystems()` will run continuously, but each run of the command should still check whether warming-up is needed _at that exact moment_.
2. Right, missed it. However, this defeats a bit the purpose of `calculateTargets()`, which should be the only function to do stuff like that.
3. The problem is that e.g. forward Amping may be desired on the state when `calculateTargets()` is called, but by the time `score()` is called, reversed Amping will be more efficient.
4. OK. Hopefully, the robot won't tip.

I hope that explains things better.

Reply:
>     3. The problem is that e.g. forward Amping may be desired on the state when `calculateTargets()` is called, but by the time `score()` is called, reversed Amping will be more efficient.
No, this situation will not happen because it means that the driver went really really close to the amp, and then backed away from it significantly.

Reply:
> this situation will not happen because it means that the driver went really really close to the amp, and then backed away from it significantly.

First, this is a situation that can happen, for example, if the coach decides to take a Note near the Amp, then has to drive to the other side of the field because another robot was blocking the robot's way on this side.
But, I was thinking of another possibility -- think of a very simple case where the robot is 5m from the Amp while its "forward" side points to the Amp when `calculateTargets()` is run, determining forward Amping is desired. Then the robot rotates 180°, meaning now reversed Amping is desired. Because the state of `isAmpingForward` doesn't change after being initially set, the robot will still run forward Amping in `score()`, meaning it will needlessly rotate and score forward, while it can score backwards.

Reply:
Can this be resolved?

Reply:
Yes. Continuing discussion in https://github.com/Galaxia5987/robot-2024/pull/55#discussion_r1486242709.